topdown/encoding.go

```diff
@@ -35,6 +35,116 @@ func builtinJSONMarshal(_ BuiltinContext, operands []*ast.Term, iter func(*ast.T
 	return iter(ast.StringTerm(string(bs)))
 }
 
+func builtinJSONMarshalWithOpts(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+
+	asJSON, err := ast.JSON(operands[0].Value)
+	if err != nil {
+		return err
+	}
+
+	indentWith := "\t"
+	prefixWith := ""
+	implicitPrettyPrint := false
+	userDeclaredExplicitPrettyPrint := false
+	shouldPrettyPrint := false
+
+	switch operands[1].Value.(type) {
+	case ast.Null:
+		// Do nothing - use default parameters
+	case ast.Object:
+		marshalOpts, err := builtins.ObjectOperand(operands[1].Value, 2)
+		if err != nil {
+			return err
+		}
+
+		for idx, k := range marshalOpts.Keys() {
+
+			val := marshalOpts.Get(k)
+
+			key, err := builtins.StringOperand(k.Value, idx)
+			if err != nil {
+				return builtins.NewOperandErr(2, "failed to stringify key %v at index %d: %v", k, idx, err)
+			}
+
+			switch key {
+
+			case "prefix":
+				switch val.Value.(type) {
+				case ast.Null:
```

Comment:
If we drop the optional values from the built-in's argument declaration, we can drop this check. Same for the other parameters.

Reply:
7990d66433ae207a393648b720b234c24cf391e9 removed the type-switches for null value handling.